File: src/main/scala/com/strings/leetcode/array/Problem_1232_checkStraightLine.scala
package com.strings.leetcode.array

object Problem_1232_checkStraightLine {

  def main(args: Array[String]): Unit = {
    val cor = Array(Array(1, 2), Array(2, 3), Array(3, 4), Array(4, 5))
    val cor2 = Array(Array(1, 2), Array(2, 2), Array(3, 2), Array(4, 2))
    val cor3 = Array(Array(1, 2), Array(2, 2), Array(3, 2), Array(4, 3))
//    println(checkStraightLine(cor))
//    println(checkStraightLine(cor2))
    println(checkStraightLine(cor3))
  }

  def checkStraightLine(coordinates: Array[Array[Int]]): Boolean = {
    if (coordinates.length == 2) true
    else {
      val flag = new Array[Double](coordinates.length - 1)
      for (a <- 1 until coordinates.length) {
        val arr: Array[Int] = coordinates(a)
        if (arr(0) - coordinates(0)(0) == 0) {
          flag(a - 1) = Int.MaxValue
        } else {
          flag(a - 1) = (arr(1).toDouble - coordinates(0)(1).toDouble) / (arr(0).toDouble - coordinates(0)(0).toDouble)
        }
      }
      println(flag.mkString(" "))
      if(flag.toSet.seq.size == 1) true
      else false
    }
  }
}

File: src/main/java/leetcode/array/Problem_1232_checkStraightLine.java
package leetcode.array;

public class Problem_1232_checkStraightLine {
    public static void main(String[] args) {
        int[][] cor = {{1,2},{2,3},{3,4}};
        int[][] cor2 = {{1,2},{2,4},{3,4}};
        System.out.println(checkStraightLine(cor));
        System.out.println(checkStraightLine(cor2));

    }


    public static boolean checkStraightLine(int[][] coordinates) {
        int size = coordinates.length;
        if(size == 2) return true;
        int x0 = coordinates[0][0],y0 = coordinates[0][1];
        double[] flag = new double[size - 1];
        for (int i = 1; i < size ; i++) {
            int[] arr_i = coordinates[i];
            if(arr_i[0] - x0 == 0) flag[i-1] = Double.MAX_VALUE;
            else  flag[i-1] =(double)(arr_i[1] - y0)/(arr_i[0] - x0);
        }

        for (int i = 1; i < flag.length ; i++) {
            if(flag[i] != flag[0]) return false;
        }

        return true;
    }
}
